### PR TITLE
virtual app commitments, messages, AppRegistry changes

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**
This PR adds commitments and protocol steps for the install-virtual-app protocol.

- adds `MixinVirtualAppSetState.sol`, adding a method `virtualAppSetState ` to the app registry that is similar to `setState`. The only difference is that the first signature, instead of signing something that allows the app state to be set to a particular value and version number, signs something that allows it to be set to anything as long as the version number is below a certain value (chosen as `65536` in the client-side code). This means that non-intermediary participants can set the app version number ("nonce") to values in the range `0, 1, ... 65536` without interacting with the intermediary, and that when it comes time to uninstall a virtual app, all participants (including the intermediary) can sign an update at version number `65537` which supercedes all previous updates
- adds `ETHVirtualAppAgreementCommitment`, an `EthereumCommitment` representing commitments to call the `ETHVirtualAppAgreement::delegateTarget` function
- adds `VirtualAppSetStateCommitment`, an `EthereumCommitment` representing commitments to call the `virtualAppSetState` function on the app registry
- adds a field `ETHVirtualAppAgreementCommitmentInstances` to the state channel model (and to firebase), mirroring the `appInstances` field

## generalizations made that affect other protocols

- `context.stateChannel` is gone, since the intermediary deals with 2 or 3 `StateChannel` instances (called the "left channel" and the "right channel"). It is replaced by a `Map<string, StateChannel>` with keys being the multisig address of a channel. This matches the way the node stores `StateChannel` instances in memory
- `ProtocolMessage.multisigAddress` is gone, for the same reason. multisig addresses have been folded into the various `<ProtocolName>Param` types. Notably, the `InstallVirtualAppParams` type has two multisig address fields

## API hacks

I changed some APIs in a very hacky way to get this working. The reason they needed changing is similar in kind to the reason the changes in the above section were made. Care needs to be taken to properly plan a good API.

- `EthereumCommitment::hashToSign()` can no longer return a single hash, since the hash that the intermediary signs is different from the one the other participants sign. The hacky solution was to pass an optional `signerIsIntermediary` flag that is only used by the install-virtual-app protocol. Obviously this is extremely ugly.
- Similarly, `EthereumCommitment:: transaction` takes an optional `intermediarySignature` field, since the intermediary signature must be placed at the front of the signatures bytestring
- the fields `{commitment, signature}{2, 3}` were added to instruction executor `Context` since the intermediary generates and signs 3 commitments
- `signature{2, 3}` was added to `ProtocolMessage` since messages contain up to 3 signatures. Note that some of these signatures are copied from other messages (i.e., the message sender "forwards" a signature they receive).
- A `VirtualAppSetStateCommitment` commitment takes two optional parameters, `hashedAppState` and `appLocalNonce`. This is because the current `EthereumCommitment` ties the concept of a "thing to sign" and a "transaction that can be done" into one obejct (these optional parameters are needed for the "transaction that can be done" but not the "thing that the intermediary needs to sign". Note, they are part of the "thing that the non-intermediary parties need to sign".) 

## other todo items

note: some of these could be done in a PR after this one as they are not technically required for the install-virtual-app protocol to work

- make `ETHVirtualAppAgreement.sol` cancellable
- bignumberify the fields `expiry` and `capitalProvided` of `ETHVirtualAppAgreementInstance` (note: this is nontrivial as these fields go into firebase, so we need to convert bignumbers to/from JSON)
- change the in-firebase representations of `appInstances` and `ETHVirtualAppAgreementInstances`; they should probably be javascript objects instead of arrays of key-value pairs
- place the target virtual app into a `StateChannel` instance; generally hook everything up so that the installation of a virtual app is writteable into firebase under some key. note that we cannot use "multisig" as the key anymore since there is no on-chain multisig.

### CODE REVIEW

File: packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    );
+
+    bytes32 digest2 = keccak256(
+      abi.encodePacked(
+        byte(0x19),
+        id,
+        req.stateHash,
+        req.nonceExpiry,
+        req.timeout,
+        byte(1)
```
Comment (snario) [inline packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol]: Why?

File: packages/machine/src/opcodes.ts [hunk 2]
```diff
@@ -1,15 +1,6 @@
...
-   * completes. Useful for other opcodes that may need to know about such state,
-   * for example, to generate the correct cf operation.
-   */
-  STATE_TRANSITION_PROPOSE,
-
-  /**
-   * Saves the new state upon completion of a protocol, using the state from
-   * STATE_TRANSITION_PROPOSE. Assumes all messages have been exchanged and
-   * the state has gone through PROPOSE and PREPARE already.
+   * Currently unused.
```
Comment (snario) [inline packages/machine/src/opcodes.ts]: I think we should keep the old docstring and add this as a note.

File: packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol (hunk 1 above)
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol]: Otherwise signatures of  the two types cant be distinguished

File: packages/machine/src/protocol/install.ts [hunk 3]
```diff
@@ -100,14 +100,14 @@ function proposeStateTransition(
     bobBalanceDecrement
   );
 
-  context.operation = constructInstallOp(
+  context.commitment = constructInstallOp(
```
Comment (snario) [inline packages/machine/src/protocol/install.ts]: These changes can be their own PR

File: packages/machine/src/instruction-executor.ts [hunk 4]
```diff
@@ -46,12 +36,15 @@ export class InstructionExecutor {
...
+  public async runUpdateProtocol(sc: StateChannel, params: UpdateParams) {
     const protocol = Protocol.Update;
     return this.runProtocol(sc, getProtocolFromName(protocol)[0], {
       params,
       protocol,
-      ...genericProtocolMessageFields(sc)
+      multisigAddress: sc.multisigAddress,
+      seq: 0,
+      fromAddress: params.initiatingAddress,
+      toAddress: params.respondingAddress
```
[the remark this answers is not in the document]
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/instruction-executor.ts]: generally yes, but the replacement for `genericProtocolMessageFields` is still a bit WIP and being figured out

File: packages/machine/src/models/state-channel.ts [hunk 5]
```diff
@@ -41,6 +45,10 @@ export type StateChannelJSON = {
   readonly multisigAddress: string;
   readonly multisigOwners: string[];
   readonly appInstances: [string, AppInstanceJson][];
+  readonly ETHVirtualAppAgreementInstances: [
+    string,
+    ETHVirtualAppAgreementJson
+  ][];
```
[the remark this answers is not in the document]
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/models/state-channel.ts]: how come `appInstances` is an array of arrays?

File: packages/machine/src/models/state-channel.ts [hunk 6]
```diff
@@ -271,7 +338,14 @@ export class StateChannel {
         }
       ),
       freeBalanceAppIndexes: Array.from(this.freeBalanceAppIndexes.entries()),
-      monotonicNumInstalledApps: this.monotonicNumInstalledApps
+      monotonicNumInstalledApps: this.monotonicNumInstalledApps,
+      ETHVirtualAppAgreementInstances: [
+        ...this.ethVirtualAppAgreementInstances.entries()
+      ].map(
```
[the remark this answers is not in the document]
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/models/state-channel.ts]: yes. I copied this from the same code in `appInstances`

File: packages/machine/src/models/state-channel.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (snario) [reply packages/machine/src/models/state-channel.ts]: that is the implementation of the ES6 [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) type

File: packages/machine/src/models/state-channel.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (snario) [reply packages/machine/src/models/state-channel.ts]: I think it just needs that for the `.filter()` method

File: packages/machine/test/unit/ethereum/install-commitment.spec.ts [hunk 7]
```diff
@@ -27,15 +28,7 @@ describe("InstallCommitment", () => {
...
-  const networkContext: NetworkContext = {
-    ETHBucket: getAddress(hexlify(randomBytes(20))),
-    StateChannelTransaction: getAddress(hexlify(randomBytes(20))),
-    MultiSend: getAddress(hexlify(randomBytes(20))),
-    NonceRegistry: getAddress(hexlify(randomBytes(20))),
-    AppRegistry: getAddress(hexlify(randomBytes(20))),
-    ETHBalanceRefund: getAddress(hexlify(randomBytes(20))),
-    ETHVirtualAppAgreement: getAddress(hexlify(randomBytes(20)))
-  };
+  const networkContext = generateRandomNetworkContext();
```
Comment (snario) [inline packages/machine/test/unit/ethereum/install-commitment.spec.ts]: This could be its own PR

File: packages/machine/src/models/state-channel.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/models/state-channel.ts]: actually, `x.entries()` has no `map` method

**Comment by cf19drofxots** [comment]
@IIIIllllIIIIllllIIIIllllIIIIllllIIIIll can you confirm this is no longer WIP?
